llama-index-core/llama_index/core/llms/structured_llm.py

```diff
@@ -0,0 +1,207 @@
+    achat_to_completion_decorator,
+    chat_to_completion_decorator,
+)
+from llama_index.core.base.query_pipeline.query import (
+    InputKeys,
+    OutputKeys,
+    QueryComponent,
+)
+
+if TYPE_CHECKING:
+    from llama_index.core.chat_engine.types import AgentChatResponse
+
+
+class StructuredLLM(LLM):
+    """
+    A structured LLM takes in an inner LLM along with a designated output class,
+    and all methods will return outputs in that structure.
+
+    """
+
+    llm: LLM
+    output_cls: Type[BaseModel] = Field(
+        ..., description="Output class for the structured LLM.", exclude=True
+    )
+
+    @classmethod
+    def class_name(cls) -> str:
+        return "structured_llm"
+
+    @property
+    def metadata(self) -> LLMMetadata:
+        return self.llm.metadata
+
+    @llm_chat_callback()
+    def chat(self, messages: Sequence[ChatMessage], **kwargs: Any) -> ChatResponse:
+        """Chat endpoint for LLM."""
+        # TODO:
+
+        # NOTE: we are wrapping existing messages in a ChatPromptTemplate to
+        # make this work with our FunctionCallingProgram, even though
+        # the messages don't technically have any variables (they are already formatted)
+
+        chat_prompt = ChatPromptTemplate(message_templates=messages)
+
+        output = self.llm.structured_predict(
+            output_cls=self.output_cls, prompt=chat_prompt, **kwargs
+        )
+        return ChatResponse(
+            message=ChatMessage(role=MessageRole.ASSISTANT, content=output.json()),
+            raw=output,
+        )
+
+    @llm_chat_callback()
+    def stream_chat(
+        self, messages: Sequence[ChatMessage], **kwargs: Any
+    ) -> ChatResponseGen:
+        raise NotImplementedError("stream_chat is not supported by default.")
+
+    @llm_completion_callback()
+    def complete(
+        self, prompt: str, formatted: bool = False, **kwargs: Any
+    ) -> CompletionResponse:
+        complete_fn = chat_to_completion_decorator(self.chat)
+        return complete_fn(prompt, **kwargs)
+
+    @llm_completion_callback()
+    def stream_complete(
+        self, prompt: str, formatted: bool = False, **kwargs: Any
+    ) -> CompletionResponseGen:
+        """Stream completion endpoint for LLM."""
+        raise NotImplementedError("stream_complete is not supported by default.")
+
+    # ===== Async Endpoints =====
+    @llm_chat_callback()
+    async def achat(
+        self,
+        messages: Sequence[ChatMessage],
+        **kwargs: Any,
+    ) -> "AgentChatResponse":
+        # NOTE: we are wrapping existing messages in a ChatPromptTemplate to
+        # make this work with our FunctionCallingProgram, even though
+        # the messages don't technically have any variables (they are already formatted)
+
+        chat_prompt = ChatPromptTemplate(message_templates=messages)
+
+        output = await self.llm.astructured_predict(
+            output_cls=self.output_cls, prompt=chat_prompt, **kwargs
```

Comment:
We probably shouldn't pass in kwargs here, since those are the prompt args (and could cause hard-to-debug issues)

Reply:
...yes you're right. fixing